harness/determined/searcher/search_runner.py

```diff
@@ -0,0 +1,243 @@
+                            exited_reason=ExitedReason._from_bindings(
+                                event.trialExitedEarly.exitedReason
+                            ),
+                        )
+                    elif event.validationCompleted:
+                        # duplicate completion accounting already performed by master
+                        logging.info(
+                            f"validationCompleted({event.validationCompleted.requestId},"
+                            f" {event.validationCompleted.metric})"
+                        )
+                        request_id = uuid.UUID(event.validationCompleted.requestId)
+                        if event.validationCompleted.metric is None:
+                            raise RuntimeError(
+                                "validationCompleted event is invalid without a metric"
+                            )
+                        operations = self.search_method.on_validation_completed(
+                            request_id,
+                            event.validationCompleted.metric,
+                        )
+                    elif event.experimentInactive:
+                        logging.info(
+                            f"experiment {self.search_method.searcher_state.experiment_id} is "
+                            f"inactive; state={event.experimentInactive.experimentState}"
+                        )
+                        experiment_is_active = False
+                        break
+                    elif event.trialProgress:
+                        logging.debug(
+                            f"trialProgress({event.trialProgress.requestId}, "
+                            f"{event.trialProgress.partialUnits})"
+                        )
+                        request_id = uuid.UUID(event.trialProgress.requestId)
+                        self.search_method.searcher_state.trial_progress[request_id] = float(
+                            event.trialProgress.partialUnits
+                        )
+                        progress = self.search_method.progress()
+                        operations = [Progress(progress)]
+                    else:
+                        raise RuntimeError(f"Unsupported event {event}")
+
+                    if skip_posting:
+                        logging.warning(f"event {event.id} has already been acknowledged by master")
+                    else:
+                        body = bindings.v1PostSearcherOperationsRequest(
+                            experimentId=self.search_method.searcher_state.experiment_id,
+                            searcherOperations=[op._to_searcher_operation() for op in operations],
+                            triggeredByEvent=event,
+                        )
+                        bindings.post_PostSearcherOperations(
+                            session,
+                            body=body,
+                            experimentId=experiment_id,
+                        )
```

Comment:
I think, since this API is designed for very advanced users, we should consider being more flexible in the programming paradigms we allow.

Right now, the rules for user code are basically: 
1. you must write event-based programming
2. it has to fit into an event loop that we write that is based on us polling the master
3. you are only allowed to emit new operations in response to master events.

But what I see from the REST api is that technically the searcher could post new operations at any time.  Which means 3. isn't a requirement we actually need to enforce on the user.  Event 1. isn't _technically_ a requirement, though I don't know how a user could do anything useful without it.  And because I assume the users of this API will be quite advanced, 2. seems undesirable to enforce.  Users who decide to come up with new events at some random time (maybe after some computationally heavy process, or due to some external event) should be able to post new events without waiting for the next event from the master, it seems like.

We should talk more about this in a meeting, I'm thinking.